mono/mini/memory-access.c

```diff
@@ -51,39 +54,51 @@ mini_emit_memset (MonoCompile *cfg, int destreg, int offset, int size, int val,
 	else
 		MONO_EMIT_NEW_ICONST (cfg, val_reg, val);
 
-	if (align < 4) {
-		/* This could be optimized further if neccesary */
-		while (size >= 1) {
-			MONO_EMIT_NEW_STORE_MEMBASE (cfg, OP_STOREI1_MEMBASE_REG, destreg, offset, val_reg);
-			offset += 1;
-			size -= 1;
-		}
-		return;
-	}	
-
-	if (!cfg->backend->no_unaligned_access && SIZEOF_REGISTER == 8) {
-		if (offset % 8) {
-			MONO_EMIT_NEW_STORE_MEMBASE (cfg, OP_STOREI4_MEMBASE_REG, destreg, offset, val_reg);
-			offset += 4;
-			size -= 4;
-		}
+	if (align < SIZEOF_VOID_P) {
+		if (align % 2 == 1)
+			goto set_1;
+		if (align % 4 == 2)
+			goto set_2;
+		if (SIZEOF_VOID_P == 8 && align % 8 == 4)
+			goto set_4;
+	}
+
+	//Unaligned offsets don't naturaly happen in the runtime, so it's ok to be conservative in how we copy
+	//We assume that input src and dest are be aligned to `align` so offset just worsen it
+	int offsets_mask = offset & 0x7; //we only care about the misalignment part
+	if (offsets_mask) {
+		if (offsets_mask % 2 == 1)
+			goto set_1;
+		if (offsets_mask % 4 == 2)
+			goto set_2;
+		if (SIZEOF_VOID_P == 8 && offsets_mask % 8 == 4)
+			goto set_4;
+	}
+
+	if (SIZEOF_REGISTER == 8) {
```

Comment:
It's mostly equivalent anyway, since the compiler will CF+DCE it on 32-bit targets. It reads nicer than preprocessor directives IMO, and we will need to do it this way anyway if we want to eventually support cross-compiling for different bitness.